### DISCUSSION TITLE
xunit.v3 support

### DISCUSSION THREAD

**Discussion by adam-knights** [reqnroll/Reqnroll#363] [Ideas]
XUnit.v3 is now available: https://xunit.net/releases/v3/1.0.0

Discussion for supporting that.

It's a package rename, so we may prefer `Reqnroll.xUnit.v3` too?

**Comment by kaylumah** [comment]
One example I have found online is keeping both versions of the package alive, like for example done by Verify
https://github.com/VerifyTests/Verify 

https://www.nuget.org/packages/Verify.XunitV3/
https://www.nuget.org/packages/Verify.Xunit/

Doing the same would not force people to migrate if they don't which to migrate.

**Reply by kaylumah** [reply]
Due note, we also have this plugin  (and perhaps others) https://github.com/reqnroll/Reqnroll/blob/main/Plugins/Reqnroll.Verify/Reqnroll.Verify.ReqnrollPlugin/Reqnroll.Verify.ReqnrollPlugin.csproj

That explicitly use the old version of xunit

**Reply by ajeckmans** [reply]
Yes and we definitely need a v3 version of that since verify has one as well

**Comment by 304NotModified** [comment]
So I think we should do the following to support xUnit 3:
- Create a new generator provider for xUnit 3 (see https://github.com/reqnroll/Reqnroll/blob/main/Reqnroll.Generator/UnitTestProvider/XUnit2TestGeneratorProvider.cs). Warning, new location, see TUnit PR #442
- Create tests
- Create an new NuGet package Reqnroll.xUnit.v3 (that depends on xUnit 3 package, I think https://www.nuget.org/packages/xunit.v3.core?)
- Update the docs
- Create Reqnroll.Verify.xUnit.v3?

Am I missing something? 

Update: updated the list above

**Reply by ajeckmans** [reply]
Keep in mind you will need an alternative for the xunit.skippablefact package as well: https://aarnott.github.io/Xunit.SkippableFact/docs/xunit-v3.html
Also the Reqnroll.Verify plugin is the only other one that also uses xunit. It would be nice if this could also gets a v3 version.

**Reply by gasparnagy** [reply]
You can also take a look at how @AdaskoTheBeAsT did the TUnit support: https://github.com/reqnroll/Reqnroll/pull/442 as additional inspiration

**Reply by 304NotModified** [reply]
> Keep in mind you will need an alternative for the xunit.skippablefact package as well: https://aarnott.github.io/Xunit.SkippableFact/docs/xunit-v3.html

But the page says, "Xunit v3 has skipping built-in, so there is no need for Xunit.SkippableFact. As part of upgrading from Xunit v2 to v3, you should remove your package reference on Xunit.SkippableFact."

So we don't need an alternative? 

Anyway, plugins/extensions is a next step IMO

**Reply by 304NotModified** [reply]
ow this is important:

> One small comment: traditionally the unit test provider classes have been added to the UnitTestProvider folder of the generator project, but this was because at that time we did not have a plugin system. In your case, you can (and you should IMHO) put your unit test provider class to the generator plugin project and add move the registration to the plugin class like we do for example [here](https://github.com/reqnroll/Reqnroll/blob/bd52d8c6cb923fe020520f1abbde08cbc8743168/Plugins/Reqnroll.ExternalData/Reqnroll.ExternalData.ReqnrollPlugin/ExternalDataGeneratorPlugin.cs?plain=1#L20).

> 
> (In principle we should have moved the NUnit/xUnit/MsTest stuff as well to their plugins, just we haven't done that yet.)

See https://github.com/reqnroll/Reqnroll/pull/442#issuecomment-2650024106

Maybe we should move first xUnit 2, then it's easier to make things consistent

**Reply by ajeckmans** [reply]
Yes, so you will need to change the code from v2 to v3 to keep that behaviour working.

**Comment by 304NotModified** [comment]
See also https://github.com/reqnroll/Reqnroll/issues/405

Maybe we should close this discussion in favor of #405

**Reply by gasparnagy** [reply]
OK. Let's continue the discussion there.
